### PR TITLE
ST25DV sensor driver implementation

### PR DESCRIPTION
**Pull request by ABOSTM**
**Summary**
ST25DV sensor driver implementation

Driver ST25DV extracted from X-NUCLEO-NFC04A1  
 *   https://github.com/stm32duino/X-NUCLEO-NFC04A1  
 *   SHA1: 73c90732098205e4fc052e509c97fd7623c30416

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,53 @@
 # ST25DV
-Arduino library to support ST25DV components
+Arduino library to support the NFC ST25DV
```
Comment (fpistm) [inline README.md]:
```suggestion
Arduino library to support the NFC ST25DV components:
 * ST25DV04K
 * ST25DV64K
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,53 @@
...
+
+## API
+
+This NFC sensor uses I2C to communicate.
+It creates the instance st25dv.
+
+It is then required to call begin API with the 2 pins to be used before accessing to the sensors, and optionally wire instance:
+    st25dv->begin(gpo_pin, lpd_pin);
```
Comment (fpistm) [inline README.md]:
```suggestion
`    st25dv->begin(gpo_pin, lpd_pin);`
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,53 @@
...
+or
+    TwoWire MyWire(SDA_PIN, SCL_PIN);
+    st25dv->begin(gpo_pin, lpd_pin, MyWire);
```
Comment (fpistm) [inline README.md]:
```suggestion
```
    TwoWire MyWire(SDA_PIN, SCL_PIN);
    st25dv->begin(gpo_pin, lpd_pin, MyWire);
```
```

File: README.md [hunk 4]
```diff
@@ -1,2 +1,53 @@
...
+
+
+It is then possible to read/write NFC URI:
+
+    int writeURI(String protocol, String uri, String info);
+    String readURI();
```
Comment (fpistm) [inline README.md]:
```suggestion
```
    int writeURI(String protocol, String uri, String info);
    String readURI();  
```
```

File: README.md [hunk 5]
```diff
@@ -1,2 +1,53 @@
...
+
+* ST25DV_SimpleWrite: This application writes a NDEF message, containg a URI record, to the tag.
+
+When the NFC module is started and ready, the message "System init done!" is displayed on the monitor window.
+Next, the tag is written with a URI.
+
+You can test this application by tapping the tag with your smartphone.
+On Android, check if NFC is activated on your smartphone.
+Put your smartphone near the tag to read it.
+The prefered Internet Browser is automatically opened with the provided URI.
```
Comment (fpistm) [inline README.md]:
```suggestion
```

File: README.md [hunk 6]
```diff
@@ -1,2 +1,53 @@
...
+
+
+## Documentation
+
+You can find the source files at  
+https://github.com/stm32duino/ST25DV
+
+The ST25DV datasheets is available at
```
Comment (fpistm) [inline README.md]:
```suggestion
The ST25DV datasheets are available at  
```

File: LICENSE.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
Comment (fpistm) [inline LICENSE.md]: Duplicated file, LICENSE already exists. To be deleted.
```suggestion
```

File: src/BSP/x_nucleo_nfc04a1_nfctag.cpp [hunk 8]
```diff
@@ -0,0 +1,297 @@
+/**
+  ******************************************************************************
+  * @file    x_nucleo_nfc04a1_nfctag.c
```
Comment (fpistm) [inline src/BSP/x_nucleo_nfc04a1_nfctag.cpp]: This file should be renamed to `nfctag.c` or `st25dv_nfctag.c` and all `x_nucleo_nfc04a1_` have to be removed.
```suggestion
  * @file    st25dv_nfctag.c
```

File: src/BSP/x_nucleo_nfc04a1_nfctag.h [hunk 9]
```diff
@@ -0,0 +1,99 @@
+/**
+  ******************************************************************************
+  * @file    x_nucleo_nfc04a1_nfctag.h
```
Comment (fpistm) [inline src/BSP/x_nucleo_nfc04a1_nfctag.h]: Ditto
```suggestion
  * @file   st25dv_nfctag.h
```

File: src/ST25DVSensor.h [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  * the "License"; You may not use this file except in compliance with the
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+#ifndef _ST25DVSENSOR_H_
+#define _ST25DVSENSOR_H_
+#include "Arduino.h"
+#include "BSP/x_nucleo_nfc04a1_nfctag.h"
```
Comment (fpistm) [inline src/ST25DVSensor.h]: Ditto

File: src/libNDEF/NDEFcommon.h [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  *
+  ******************************************************************************
+  */
+
+/* Define to prevent recursive inclusion -------------------------------------*/
+#ifndef __COMMON_H
+#define __COMMON_H
+
+/* Includes ------------------------------------------------------------------*/
+//#include "x_nucleo_nfc04a1_nfctag.h"
```
Comment (fpistm) [inline src/libNDEF/NDEFcommon.h]:
```suggestion
```

File: src/libNDEF/NDEFcommon.h [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+#include <string.h>
+#include "stdint.h"
+
+/* Exported types ------------------------------------------------------------*/
+//typedef uint8_t boolean;
```
Comment (fpistm) [inline src/libNDEF/NDEFcommon.h]:
```suggestion
```

File: src/libNDEF/NDEFcommon.h [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+#define MIN(x, y) (((x) < (y)) ? (x) : (y))
+#endif
+
+/*
+#undef FAIL
+#define FAIL 0
+
+#undef PASS
+#define PASS !FAIL
+*/
```
Comment (fpistm) [inline src/libNDEF/NDEFcommon.h]:
```suggestion
```

File: README.md (hunk 1 above)
Comment (ABOSTM) [reply README.md]: Done

File: README.md (hunk 2 above)
Comment (ABOSTM) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (ABOSTM) [reply README.md]: Done

File: README.md (hunk 4 above)
Comment (ABOSTM) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (ABOSTM) [reply README.md]: Done

File: README.md (hunk 6 above)
Comment (ABOSTM) [reply README.md]: Done

File: LICENSE.md (hunk 7 above)
Comment (ABOSTM) [reply LICENSE.md]: Done

File: src/BSP/x_nucleo_nfc04a1_nfctag.cpp (hunk 8 above)
Comment (ABOSTM) [reply src/BSP/x_nucleo_nfc04a1_nfctag.cpp]: Done

File: src/BSP/x_nucleo_nfc04a1_nfctag.h (hunk 9 above)
Comment (ABOSTM) [reply src/BSP/x_nucleo_nfc04a1_nfctag.h]: Done

File: src/ST25DVSensor.h (hunk 10 above)
Comment (ABOSTM) [reply src/ST25DVSensor.h]: Done

File: src/libNDEF/NDEFcommon.h (hunk 11 above)
Comment (ABOSTM) [reply src/libNDEF/NDEFcommon.h]: Done

File: src/libNDEF/NDEFcommon.h (hunk 12 above)
Comment (ABOSTM) [reply src/libNDEF/NDEFcommon.h]: I propose to keep it like this as this is something I didn't change (I didn't touch to libNDEF files)

File: src/libNDEF/NDEFcommon.h (hunk 13 above)
Comment (ABOSTM) [reply src/libNDEF/NDEFcommon.h]: I propose to keep it like this as this is something I didn't change (I didn't touch to libNDEF files)

**Comment by cparata** [comment]
Hi @ABOSTM ,
I tested the new 2 libraries and they are functional also with X-NUCLEO-NFC04A1. I'm just wondering if this is a good occasion to remove some warnings that are inside the libraries. Some of them should be quite simple to remove. What about that?
Best Regards,
Carlo

**Pull request merged**